### PR TITLE
get price from inverse pair if can't get direct pair

### PR DESCRIPTION
**Pull request by nguyenbatam**
1 . get price from direct pair in smart contract
2 . get price from inverse pair in smart contract
3.  get price from direct pair in tomox
4 . get price from inverse pair in tomox
5 . get price from pair collateral/tomo and lending/tomo or inverse ....

### CODE REVIEW

File: tomoxlending/order_processor.go [hunk 1]
```diff
@@ -861,62 +861,58 @@ func (l *Lending) GetCollateralPrices(header *types.Header, chain consensus.Chai
...
+	if err != nil || lendingTokenDecimal == nil || lendingTokenDecimal.Sign() == 0 {
+		return nil, nil, err
+	}
+	collateralTokenDecimal, err := l.tomox.GetTokenDecimal(chain, statedb, common.Address{}, collateralToken)
+	log.Debug("GetTokenDecimal", "collateralToken", collateralToken, "err", err)
+	if err != nil || collateralTokenDecimal == nil || collateralTokenDecimal.Sign() == 0 {
+		return nil, nil, err
+	}
+	var collateralPrice *big.Int
+	inverseCollateralPriceFromContract, updatedBlock := lendingstate.GetCollateralPrice(statedb, collateralToken, lendingToken)
```
Comment (thanhnguyennguyen) [inline tomoxlending/order_processor.go]: inversePair should be lendingToken/collateral

File: tomoxlending/order_processor.go [hunk 2]
```diff
@@ -861,62 +861,58 @@ func (l *Lending) GetCollateralPrices(header *types.Header, chain consensus.Chai
...
+		return nil, nil, err
 	}
-	lendingTokenDecimal, err := l.tomox.GetTokenDecimal(chain, statedb, common.Address{}, lendingToken)
-	log.Debug("GetTokenDecimal", "lendingToken", lendingToken, "err", err)
+	if lastInverseAveragePrice != nil && lastInverseAveragePrice.Sign() > 0 {
+		log.Debug("Getting collateral/lending from inverse pair in tomox", "lendToken", lendingToken.Hex(), "collateralToken", collateralToken.Hex(), "price", lastAveragePrice)
+		collateralPrice = new(big.Int).Mul(lastInverseAveragePrice, collateralTokenDecimal)
+		collateralPrice = new(big.Int).Div(collateralPrice, lendingTokenDecimal)
+		return lendTokenTOMOPrice, collateralPrice, nil
+	}
```
Comment (thanhnguyennguyen) [inline tomoxlending/order_processor.go]: not necessary 
inversePrice is calculated in  https://github.com/tomochain/tomochain/pull/255/files#diff-bb070402da3cf4811b1d48bdf02391e6R896

File: tomoxlending/order_processor.go [hunk 3]
```diff
@@ -861,62 +861,58 @@ func (l *Lending) GetCollateralPrices(header *types.Header, chain consensus.Chai
...
+	if err != nil || collateralTokenDecimal == nil || collateralTokenDecimal.Sign() == 0 {
+		return nil, nil, err
+	}
+	var collateralPrice *big.Int
+	inverseCollateralPriceFromContract, updatedBlock := lendingstate.GetCollateralPrice(statedb, lendingToken, collateralToken)
+	inverseCollateralPriceUpdatedFromContract := updatedBlock.Uint64()/chain.Config().Posv.Epoch == header.Number.Uint64()/chain.Config().Posv.Epoch
+	if inverseCollateralPriceUpdatedFromContract {
+		log.Debug("Getting lending/collateral token price from contract", "price", inverseCollateralPriceFromContract)
+		collateralPrice = new(big.Int).Mul(inverseCollateralPriceFromContract, collateralTokenDecimal)
+		collateralPrice = new(big.Int).Div(collateralPrice, lendingTokenDecimal)
```
Comment (thanhnguyennguyen) [inline tomoxlending/order_processor.go]: check this formula
I think it should be
```
collateralPrice = lendTokenDecimal * collateralTokenDecimal / inverseCollateralPriceFromContract
```

similar to https://github.com/tomochain/tomochain/blob/561213304a4be971456a113c1def81a81d3d4243/tomoxlending/order_processor.go#L845

Comment (nguyenbatam) [reply tomoxlending/order_processor.go]: updated

File: tomoxlending/order_processor.go (hunk 2 above)
Comment (nguyenbatam) [reply tomoxlending/order_processor.go]: if not found inverse price from smart contract try get it from tomox

Comment (nguyenbatam) [reply tomoxlending/order_processor.go]: removed

File: tomoxlending/order_processor.go [hunk 4]
```diff
@@ -928,10 +913,43 @@ func (l *Lending) GetCollateralPrices(header *types.Header, chain consensus.Chai
...
+		log.Debug("Getting token/TOMO price from contract", "price", tokenTOMOPriceFromContract)
+		return tokenTOMOPriceFromContract, nil
+	} else {
+		tomoTokenPriceFromContract, updatedBlock := lendingstate.GetCollateralPrice(statedb, common.HexToAddress(common.TomoNativeAddress), token)
+		tomoTokenPriceUpdatedFromContract := updatedBlock.Uint64()/chain.Config().Posv.Epoch == header.Number.Uint64()/chain.Config().Posv.Epoch
+		if tomoTokenPriceUpdatedFromContract && tomoTokenPriceFromContract != nil && tomoTokenPriceFromContract.Sign() > 0 {
+			// getting lendToken price from contract first
+			// otherwise, getting from tomox lendToken/TOMO
+			log.Debug("Getting TOMO/token from contract", "price", tomoTokenPriceFromContract)
+			return new(big.Int).Div(common.BasePrice, tokenTOMOPriceFromContract), nil
```
Comment (thanhnguyennguyen) [inline tomoxlending/order_processor.go]: common.BasePrice * tokenDecimal / tokenTOMOPriceFromContract ?

Comment (nguyenbatam) [reply tomoxlending/order_processor.go]: updated

**Review by thanhnguyennguyen — APPROVED** [review]
resolve conflict

**Review by thanhnguyennguyen — APPROVED** [review]

**Review by thanhson1085 — APPROVED** [review]

**Pull request merged**